From 254d04d4bc8095e8aa8933d9d2b3cb7f06bcac89 Mon Sep 17 00:00:00 2001
From: Colin <colin@uninsane.org>
Date: Fri, 14 Aug 2026 07:23:51 +0000
Subject: [PATCH] unifont: fix cross compilation

fixes `nix-build -A pkgsCross.aarch64-multiplatform.unifont`.

unifont uses binaries produced early in the build to perform the later
build steps. that would mean invoking binaries built for the wrong
architecture when cross-compiling, but they provide a `BINDIR=...`
Makefile option for us to provide the build-time binaries in this case.
---
 pkgs/by-name/un/unifont/package.nix | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/pkgs/by-name/un/unifont/package.nix b/pkgs/by-name/un/unifont/package.nix
index 04323ca1cb2912f0c34e7c33f372aae84ec74b0b..f008da06a4923156fe5d488755512ff2819763f6 100644
--- a/pkgs/by-name/un/unifont/package.nix
+++ b/pkgs/by-name/un/unifont/package.nix
@@ -1,6 +1,8 @@
 {
   lib,
   stdenv,
+  bashNonInteractive,
+  buildPackages,
   fetchurl,
   perl,
   bdftopcf,
@@ -25,6 +27,14 @@ stdenv.mkDerivation (finalAttrs: {
   postPatch = ''
     rm -r font/precompiled
     patchShebangs ./src
+    ${lib.optionalString (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
+      substituteInPlace Makefile --replace-fail \
+        'bin/unigenwidth ' \
+        '$(BINDIR)/unigenwidth '
+      substituteInPlace font/Makefile --replace-fail \
+        '"BINDIR:../../../$(BINDIR)/"' \
+        '"BINDIR:$(BINDIR)/"'
+    ''}
   '';
 
   nativeBuildInputs = [
@@ -35,10 +45,16 @@ stdenv.mkDerivation (finalAttrs: {
   ];
 
   buildInputs = [
+    bashNonInteractive
     perlenv
   ];
 
-  makeFlags = [ "PREFIX=${placeholder "out"}" ];
+  makeFlags = [
+    "PREFIX=${placeholder "out"}"
+  ]
+  ++ lib.optionals (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) [
+    "BINDIR=${buildPackages.unifont.bin}/bin"
+  ];
 
   buildFlags = [ "BUILDFONT=1" ];
 
@@ -47,6 +63,10 @@ stdenv.mkDerivation (finalAttrs: {
     moveToOutput share/unifont "$doc"
   '';
 
+  postFixup = ''
+    patchShebangs --host --update "$bin"
+  '';
+
   outputs = [
     "out"
     "bin"
